math/src/main/java/smile/math/matrix/SparseMatrix.java

```diff
@@ -175,11 +180,204 @@ public int size() {
+    public Stream<SparseMatrixEntry> nonzeros(int startColumn, int endColumn) {
+        return StreamSupport.stream(new SparseMatrixSpliterator(this, startColumn, endColumn), false);
+    }
+
+    /**
+     * Provides a spliterator for access to a sparse matrix in column major order.
+     * <p>
+     * This is exposed to facilitate lower level access to the stream API for a matrix.
+     */
+    public static class SparseMatrixSpliterator implements Spliterator<SparseMatrixEntry> {
+        private final SparseMatrix m;
+        private int col;      // current colum, advanced on split or traversal
+        private int index;    // current element within column
+        private final int fence; // one past the last column to process
+
+        SparseMatrixSpliterator(SparseMatrix matrix, int col, int fence) {
+            this.m = matrix;
+            this.col = col;
+            this.index = m.colIndex[col];
+            this.fence = fence;
+        }
+
+        public void forEachRemaining(Consumer<? super SparseMatrixEntry> action) {
+            for (; col < fence; col++) {
+                for (; index < m.colIndex[col + 1]; index++) {
+                    action.accept(new SparseMatrixEntry(m.rowIndex[index], col, m.x[index], index, m.x));
+                }
+            }
+        }
+
+        public boolean tryAdvance(Consumer<? super SparseMatrixEntry> action) {
+            if (col < fence) {
+                while (col < fence && index >= m.colIndex[col + 1]) {
+                    col++;
+                }
+                if (col < fence) {
+                    action.accept(new SparseMatrixEntry(m.rowIndex[index], col, m.x[index], index, m.x));
+                    index++;
+                    return true;
+                } else {
+                    return false;
+                }
+            } else {
+                // cannot advance
+                return false;
+            }
+        }
+
+        public Spliterator<SparseMatrixEntry> trySplit() {
+            int lo = col; // divide range in half, assuming roughly constant size of all columns
+            int mid = ((lo + fence) >>> 1) & ~1; // force midpoint to be even
+            if (lo < mid) { // split out left half
+                col = mid; // reset this Spliterator's origin
+                return new SparseMatrixSpliterator(m, lo, mid);
+            } else {
+                // too small to split
+                return null;
+            }
+        }
+
+        public long estimateSize() {
+            return (long) (m.colIndex[fence] - m.colIndex[col]);
+        }
+
+        public int characteristics() {
+            return ORDERED | SIZED | IMMUTABLE | SUBSIZED;
+        }
+    }
+
+    /**
+     * Encapsulates important information about an entry in a matrix for use
+     * in streaming, including a string that leads back to the original cell
+     * so that in-place updates are possible.
+     */
+    public static class SparseMatrixEntry {
```

Comment:
Since it is an inner class, we can just call it `Entry`.

Reply:
Done.